### PR TITLE
Connector salesforce 80

### PR DESCRIPTION
**Pull request by nbessi**

### CODE REVIEW

File: connector_salesforce/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+{'name': 'Salesforce Connector',
+ 'version': '0.1',
+ 'author': 'Camptocamp',
```
Comment (yvaucher) [inline connector_salesforce/__openerp__.py]: OCA is missing

File: connector_salesforce/controllers/main.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+        :param req: WSGI request
+        :return: success message or raise an error
+        :rtype: str
+        """
+        code = req.params.get('code')
+        state = req.params.get('state')
+        if not all([code, state]):
+            raise ValueError(
+                'Authorization process went wrong '
```
Comment (yvaucher) [inline connector_salesforce/controllers/main.py:49]: Shouldn't it be translatable?

Some are and other are not in odoo code

File: connector_salesforce/salesforce_account/importer.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        data['country_id'] = country_id
+        state_id = self._state_id(record,
+                                  'ShippingState',
+                                  'ShippingCountryCode')
+        data['state_id'] = state_id
+        return data
+
+    def map_shipping_address(self, record, binding=None):
+        """Manage the Salesforce account shipping address
+        If no shipping address exist in Odoo it is created.
```
Comment (yvaucher) [inline connector_salesforce/salesforce_account/importer.py]: s/exist/exists

File: connector_salesforce/salesforce_account/importer.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                self.session.env['res.partner'].write(
+                    [shipp_id],
+                    self._prepare_shipp_addresse_data(record, current_partner)
+                )
+            else:
+                shipp_id = self.session.env['res.partner'].create(
+                    self._prepare_shipp_addresse_data(record, current_partner)
+                ).id
+        else:
+            shipp_id = False
```
Comment (yvaucher) [inline connector_salesforce/salesforce_account/importer.py]: This line can be removed, `shipp_id` is already set to `False` at line 120

File: connector_salesforce/salesforce_account/importer.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    def active(self, record):
+        return {'active': True}
+
+    @mapping
+    def property_product_pricelist(self, record):
+        """Map property pricelist using current backend setting"""
+        currency_id = self.get_currency_id(record)
+        mapping = {rec.currency_id.id: rec.pricelist_version_id.id
+                   for rec in self.backend_record.sf_entry_mapping_ids}
+        price_list_version_id = mapping.get(currency_id)
```
Comment (yvaucher) [inline connector_salesforce/salesforce_account/importer.py:176]: Why do you create a whole dict to get only a single value, wouldn't a for loop with a break be better?

File: connector_salesforce/salesforce_account/importer.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+            raise MappingError(
+                'No pricelist version configuration done for '
+                'currency %s and backend %s' % (
+                    record.get('CurrencyIsoCode'),
+                    self.backend_record.name
+                )
+            )
+        pl_model = 'product.pricelist.version'
+        price_list_version_record = self.session.env[pl_model].browse(
+            price_list_version_id
```
Comment (yvaucher) [inline connector_salesforce/salesforce_account/importer.py:187]: Is this browse necessary? can't we get already the record from `sf_entry_mapping_ids` ?

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+    In order to use it you have to add a remote application in Salesforce
+    and enable Oauth login.
+
+    The created remote access app must have following parameters:
+
+    Permitted Users -->	All users may self-authorize
+    Callback URL --> public_odoo_url/salesforce/oauth
+
+
+    Once done you have to manage your app ensure the
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py]: **to** ensure ?

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+    `Refresh token is valid until revoked` parameter is set.
+
+    The following authentication method are still exprimental
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py]: s/are/is

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+            ('pwd_token', 'Based on User, Password, Token'),
+            ('oauth2', 'OAuth 2'),
+            ('ip_filtering', 'Based on IP Filter and OrganizationId')
+        ],
+        string='Authentication Method',
+        default='oauth2',
+    )
+
+    name = fields.Char(
+        'Name',
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py:106]: You don't need this string. By default string is equal ta Capitalized field name.

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+    )
+
+    @api.model
+    def _enforce_param(self, param_name):
+        """Ensure configuration parameter is set on backend record
+
+        :param param_name: name of parameter to validate
+        :type param_name: str
+
+        :return: True is parameter is set or raise an exception
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py]: s/is/if

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 11]
```diff
@@ -0,0 +1,407 @@
...
+        :rtype: bool
+        """
+        self.ensure_one()
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py:169]: With `api.model` I don't think you should use `ensure_one()`

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+        if not getattr(self, param_name, None):
+            f_model = self.env['ir.model.fields']
+            field = f_model.search(
+                [('model', '=', self._name),
+                 ('name', '=', param_name)],
+            )
+            if len(field) == 1:
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py:176]: What about a search with limit = 1 ? Aren't field uniq by (model, name) ?

File: connector_salesforce/salesforce_backend/backend_model.py [hunk 13]
```diff
@@ -0,0 +1,407 @@
...
+                  'current authentication method') % field_name
+            )
+        return True
+
+    def _enforce_url(self):
+        """Predicate hook to see if URL must be enforced
+        when validating configuration"""
+        return True
+
+    def _validate_configuration(self):
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/backend_model.py:192]: An `@api.multi` would be polite

File: connector_salesforce/salesforce_backend/view/backend_model_view.xml [hunk 14]
```diff
@@ -0,0 +1,141 @@
+<?xml version="1.0" encoding="utf-8"?>
+<openerp>
+  <data>
+    <record id="connector_salesforce_backend_form" model="ir.ui.view">
+      <field name="name">Salesforce backend form</field>
+      <field name="model">connector.salesforce.backend</field>
+      <field name="arch" type="xml">
+        <form version="7.0" string="Salesforce Backend">
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/view/backend_model_view.xml]: You can remove `version="7.0"`

File: connector_salesforce/salesforce_backend/view/backend_model_view.xml [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+                          string="Refresh token"
+                          type="object"
+                          attrs="{'invisible': [('authentication_method', '!=', 'oauth2')]}"/>
+                </div>
+                <div attrs="{'invisible': [('authentication_method', '!=', 'oauth2')]}">
+                  <separator string="Oauth2 Setup Instructions"/>
+                  <p>In order to use Oauth2 you have to add a remote application in Salesforce
+                  and enable Oauth login.</p>
+
+                  <p> The created remote access app must have following parameters:
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/view/backend_model_view.xml]: s/must have/must have the

File: connector_salesforce/salesforce_backend/view/backend_model_view.xml [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+          </sheet>
+        </form>
+      </field>
+    </record>
+
+    <record id="connector_salesforce_list" model="ir.ui.view">
+      <field name="name">Connector Salesforce list view</field>
+      <field name="model">connector.salesforce.backend</field>
+      <field name="arch" type="xml">
+        <tree version="7.0" string="Salesforce Backend">
```
Comment (yvaucher) [inline connector_salesforce/salesforce_backend/view/backend_model_view.xml]: Another `version="7.0"` to nuke

File: connector_salesforce/salesforce_contact/common.py [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+class SalesforceContact(models.Model):
+    _inherit = 'salesforce.binding'
+    _inherits = {'res.partner': 'openerp_id'}
+    _name = 'connector.salesforce.contact'
+    _description = 'Import SF Contact into res.partner model'
+
+    openerp_id = fields.Many2one('res.partner',
+                                 string='Partner',
+                                 required=True,
+                                 select=True,
```
Comment (yvaucher) [inline connector_salesforce/salesforce_contact/common.py]: select might have no effect and provide no deprecated warning, please prefer `index=True`

File: connector_salesforce/salesforce_opportunity/common.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+class SalesforceOpportunity(models.Model):
+    _inherit = 'salesforce.binding'
+    _inherits = {'sale.order': 'openerp_id'}
+    _name = 'connector.salesforce.opportunity'
+    _description = 'Import SF Opportunity into sale.order model'
+
+    openerp_id = fields.Many2one('sale.order',
+                                 string='Sale Order Line',
+                                 required=True,
+                                 select=True,
```
Comment (yvaucher) [inline connector_salesforce/salesforce_opportunity/common.py]: index

File: connector_salesforce/salesforce_opportunity/common.py [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+class SalesforceOpportunityLineItem(models.Model):
+    _inherit = 'salesforce.binding'
+    _inherits = {'sale.order.line': 'openerp_id'}
+    _name = 'connector.salesforce.opportunity.line.item'
+    _description = 'Import SF Opportunity line item into sale.order model'
+
+    openerp_id = fields.Many2one('sale.order.line',
+                                 string='Sale Order Line',
+                                 required=True,
+                                 select=True,
```
Comment (yvaucher) [inline connector_salesforce/salesforce_opportunity/common.py]: index

File: connector_salesforce/controllers/main.py (hunk 2 above)
Comment (nbessi) [inline connector_salesforce/controllers/main.py:49]: It is a Value error that is a technical error. For me it should not be translated

File: connector_salesforce/salesforce_account/importer.py (hunk 5 above)
Comment (nbessi) [inline connector_salesforce/salesforce_account/importer.py:176]: I find it more readable, but that True we could do a loop. It is just a matter of taste as once you browse one record in relation the whole set in put in cache this does not make a lot of impact on performance.

File: connector_salesforce/salesforce_account/importer.py (hunk 6 above)
Comment (nbessi) [inline connector_salesforce/salesforce_account/importer.py:187]: No in sf_entry_mapping_ids we get a dedicated configuration model

File: connector_salesforce/salesforce_backend/backend_model.py (hunk 9 above)
Comment (nbessi) [inline connector_salesforce/salesforce_backend/backend_model.py:106]: I like to be coherent, but that just a matter of taste

File: connector_salesforce/salesforce_backend/backend_model.py (hunk 11 above)
Comment (nbessi) [inline connector_salesforce/salesforce_backend/backend_model.py:169]: First I used api.one but it had nasty side effect. That why I fallback on this strategy

File: connector_salesforce/salesforce_backend/backend_model.py (hunk 12 above)
Comment (nbessi) [inline connector_salesforce/salesforce_backend/backend_model.py:176]: I can add a limit = 1 but the test should still be done as some fields are yet not created in ir.model.fields.table. There is a bug open on this. This issue also impact geoengine

Comment (nbessi) [inline connector_salesforce/salesforce_backend/backend_model.py:176]: Fun fact I just check in DB and in ir_models.py there is no constraints that ensure unicity...

**Comment by nbessi** [comment]
Add requested fixes

**Comment by gurneyalex** [comment]
overriden by #5

**Pull request closed without merging**